src/components/Cart.js

```diff
@@ -0,0 +1,157 @@
+import React from 'react';
+import { useSelector, useDispatch } from 'react-redux';
+import {
+  List,
+  ListItem,
+  ListItemText,
+  Button,
+  Typography,
+  Container,
+  Avatar,
+} from '@mui/material';
+import { reduceItem, addItem, clearCart } from '../redux/shoppingCart';
+import axiosInstance from './../apis/axiosClient';
+import { useNavigate } from 'react-router-dom';
+
+const CartItems = () => {
+  const navigate = useNavigate();
+  const dispatch = useDispatch();
+  //function to round number to 2 decimal digits
+  const roundedNumber = (num) => Math.round((num + Number.EPSILON) * 100) / 100;
+  // values come from Redux store
+  const itemsInCart = useSelector((state) => state.cart);
+  const subtotal = roundedNumber(useSelector((state) => state.cart.subtotal)); //total price of all items
+  const tax = useSelector((state) => state.cart.tax);
+  const total = roundedNumber(useSelector((state) => state.cart.totalAmount)); //total amount after tax
+
+  const handleMoveToWishlist = async (album) => {
+    dispatch(
+      reduceItem({ album, quantity: itemsInCart.items[album.id].quantity })
+    ); // Reduce the quantity of the item in the cart first
+    const response = await axiosInstance.post('/wishlist'); // Make the call to create or get the wishlist
+    const wishlistId = response.data.wishlist._id;
+    await axiosInstance.patch(`/wishlist/${wishlistId}/add_album/${album.id}`); // Make the call to add the album to the wishlist
+    console.log('Album successfully added to wishlist');
+  };
+
+  const handleAdd = (album) => {
+    dispatch(addItem({ album, quantity: 1 }));
+  };
+
+  const handleReduce = (album) => {
+    dispatch(reduceItem({ album, quantity: 1 }));
+  };
+
+  const handleRemove = (album) => {
+    dispatch(
+      reduceItem({ album, quantity: itemsInCart.items[album.id].quantity })
+    );
+  };
+
+  const handleClearCart = () => {
+    dispatch(clearCart());
+  };
+
+  //64fd67815fca88489ab99564
+  const handleCheckout = () => {
+    // Create the order data object
+    const orderData = {
+      orderItems: Object.values(itemsInCart.items).map((item) => ({
+        album: item.album.id,
```

Comment:
Since basically all of this logic outside of the render (return value) is exactly the same as the logic in `src/components/OrderSidebar.js`, this seems like a good opportunity to extract this out so it can be reused. 

One way to approahc it would be to make each of these function calls a standalone function (eg: in a `utils/helper.js`) file, that can be called from both components. One difficulty we would run into though, is that these functions need to know about react/redux to work because they use `navigate`, `dispatch` , and `useSelector` functions. We _could_ pass in the needed values, like this:

```javascript
# utils/helpers.js
const hanldeMoveToWishlist = async (album, dispatch) => {
  ...
}

# Cart.js
import {handleMoveToWishlist} from 'src/utils/helpers.js'

const CartItems = () => {
  ...
  const dispatch = useDispatch();
  
  return {
    ...
    <Button onClick={() => handleMoveToWishlist(item.album, dispatch)}>
  }
}
```

but it's a little bit ugly. 

A better approach is if we want to make helper functions that also need to know about other hooks, is to make our own custom hooks: 

[https://www.freecodecamp.org/news/how-to-create-react-hooks/](https://www.freecodecamp.org/news/how-to-create-react-hooks/)

[https://react.dev/learn/reusing-logic-with-custom-hooks](https://react.dev/learn/reusing-logic-with-custom-hooks)

In this case, it could look something like:

```javascript
# src/redux/customHooks/useCart.js
export default function useCart() {
    const navigate = useNavigate();
    const dispatch = useDispatch();
    ...<copy all the shared code between Cart.js and OrderSidebar.js>...
    
    const variables = { itemsInCart, itemValues: Object.values(itemsInCart.items), subtotal, tax, total};
    const functions = { handleMoveToWishlist, handleAdd, handleReduce, handleRemove, handleClearCart, handleCheckout };

    return [variables, functions];
}

# in both Cart.js and OrderSidebar.js, the shared code can now be simplified to something like:
const CartItems() => {
  const [{itemValues, subtotal, tax, total}, { handleMoveToWishlist, handleAdd, handleReduce, handleRemove, handleClearCart, handleCheckout }] = useCart();
  
  return (...render code...)
}
```

Reply:
@akosasante thank you so much for explanation and an article, Akos, I learned new info and now refactored the code, so from now on we can reuse this custom hook useCart in other parts of code when necessary, I will commit now